pkg/manager/manager.go

```diff
@@ -0,0 +1,183 @@
+package manager
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/deckhouse/d8-lint/pkg/config"
+	"github.com/deckhouse/d8-lint/pkg/errors"
+	"github.com/deckhouse/d8-lint/pkg/linters/copyright"
+	no_cyrillic "github.com/deckhouse/d8-lint/pkg/linters/no-cyrillic"
+	"github.com/deckhouse/d8-lint/pkg/linters/openapi"
+	"github.com/deckhouse/d8-lint/pkg/linters/probes"
+	"github.com/deckhouse/d8-lint/pkg/logger"
+	"github.com/deckhouse/d8-lint/pkg/module"
+)
+
+const (
+	ChartConfigFilename = "Chart.yaml"
+)
+
+type Manager struct {
+	cfg     *config.Config
+	Linters LinterList
+	Modules []*module.Module
+
+	lintersMap map[string]Linter
+}
+
+func NewManager(dirs []string, cfg *config.Config) *Manager {
+	m := &Manager{
+		cfg: cfg,
+	}
+
+	// fill all linters
+	m.Linters = []Linter{
+		openapi.New(&cfg.LintersSettings.OpenAPI),
+		no_cyrillic.New(&cfg.LintersSettings.NoCyrillic),
+		copyright.New(&cfg.LintersSettings.Copyright),
+		probes.New(&cfg.LintersSettings.Probes),
+	}
+
+	m.lintersMap = make(map[string]Linter)
+	for _, linter := range m.Linters {
+		m.lintersMap[strings.ToLower(linter.Name())] = linter
+	}
+
+	// filter linters from config file
+	m.Linters = m.getEnabledLinters()
+
+	var paths []string
+
+	for i := range dirs {
+		dir, err := homedir.Expand(dirs[i])
+		if err != nil {
+			logger.ErrorF("Failed to expand home dir: %v", err)
+			continue
+		}
+		result, err := getModulePaths(dir)
+		if err != nil {
+			logger.ErrorF("Error getting module paths: %v", err)
+		}
+		paths = append(paths, result...)
+	}
+
+	for i := range paths {
+		m.Modules = append(m.Modules, module.NewModule(paths[i]))
+	}
+
+	return m
+}
+
+const (
+	modulesLimit = 10
+)
+
+func (m *Manager) Run() errors.LintRuleErrorsList {
+	result := errors.LintRuleErrorsList{}
+
+	for i := range m.Linters {
+		var g errgroup.Group
+		g.SetLimit(modulesLimit)
+		sm := sync.Mutex{}
+		for j := range m.Modules {
+			g.Go(func() error {
+				errs, err := m.Linters[i].Run(m.Modules[j])
+				if err != nil {
+					logger.WarnF("Error running linter `%s`: %s\n", m.Linters[i].Name(), err)
+					return err
+				}
+				if errs.ConvertToError() != nil {
+					sm.Lock()
+					result.Merge(errs)
+					sm.Unlock()
+				}
+				return nil
+			})
+		}
+		_ = g.Wait()
+	}
+
+	return result
+}
+
+func isExistsOnFilesystem(parts ...string) bool {
+	_, err := os.Stat(filepath.Join(parts...))
+	return err == nil
+}
+
+// getModulePaths returns all paths with Chart.yaml
```

Comment:
I see three cases, when we can figure out it's a module
- directory contains `module.yaml`
- directory contains `Chart.yaml` + `/hooks` directory
- directory contains `Chart.yaml` + `/images` directory


otherwise https://github.com/deckhouse/deckhouse/tree/3db4ef69a8cdf2db5798d05ebb2e01de1f6e2774/helm_lib will be treated as a module

Reply:
As far as I understand, the second and third cases are covered by our logic, that is, we are looking for directories containing the Chart.yaml file. Which directories are nearby does not matter much. But can the module.yaml file be located alone, without Chart.yaml? If not, then our check for the module directory will be enough.

Reply:
it can, Chart.yaml is optional for our modules